diff --git spec/spec_helper.rb spec/spec_helper.rb
index 5574c4ae3..ad9e503ea 100644
--- spec/spec_helper.rb
+++ spec/spec_helper.rb
@@ -51,6 +51,12 @@ ENV["THOR_COLUMNS"] = "10000"

 Spec::CodeClimate.setup

+module Gem
+  def self.ruby= ruby
+    @ruby = ruby
+  end
+end
+
 RSpec.configure do |config|
   config.include Spec::Builders
   config.include Spec::Helpers
@@ -102,9 +108,15 @@ RSpec.configure do |config|
     system_gems []
     in_app_root
     @all_output = String.new
+    @orig_ruby = if ENV['BUNDLE_RUBY']
+                   ruby = Gem.ruby
+                   Gem.ruby = ENV['BUNDLE_RUBY']
+                   ruby
+                 end
   end

   config.after :each do |example|
+    Gem.ruby = @orig_ruby
     @all_output.strip!
     if example.exception && !@all_output.empty?
       warn @all_output unless config.formatters.grep(RSpec::Core::Formatters::DocumentationFormatter).empty?
